### ISSUE TITLE
[Bug] Can't create user due to Host ... is not allowed to connect to this MariaDB server

### ISSUE THREAD

**Issue by mdbooth** [mariadb-operator/mariadb-operator#89]
**Describe the bug**

mariadb-operator create a user due to `Error 1130: Host '10.244.1.39' is not allowed to connect to this MariaDB server`. E.g.:

```
{"level":"error","ts":1682341296.5456517,"msg":"Reconciler error","controller":"user","controllerGroup":"mariadb.mmontes.io","controllerKind":"User","user":{"name":"photoprism","namespace":"mariadb"},"namespace":"mariadb","name":"photoprism","reconcileID":"9854f729-1653-428c-9f38-c431fd603f65","error":"error reconciling in TemplateReconciler: error creating MariaDB client: 1 error occurred:\n\t* Error 1130: Host '10.244.1.39' is not allowed to connect to this MariaDB server\n\n","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:273\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.12.3/pkg/internal/controller/controller.go:234"}
```

I can presumably log in to MariaDB to fix this manually.

**Expected behaviour**
mariadb-operator creates a user.

**Steps to reproduce the bug**

Install the 0.11.0 helm chart with no customisations, then apply the yaml below.

**Additional context**

Deployed CRs:
```yaml
apiVersion: mariadb.mmontes.io/v1alpha1
kind: MariaDB
metadata:
  name: mariadb
spec:
  rootPasswordSecretKeyRef:
    name: passwords
    key: root-password

  image:
    repository: mariadb
    tag: "10.7.4" # "10.11.2"
    pullPolicy: IfNotPresent

  port: 3306

  volumeClaimTemplate:
    resources:
      requests:
        storage: 10Gi
    storageClassName: rook-ceph-block
    accessModes:
      - ReadWriteOnce

  env:
    - name: TZ
      value: UTC
---
apiVersion: mariadb.mmontes.io/v1alpha1
kind: User
metadata:
  name: photoprism
spec:
  mariaDbRef:
    name: mariadb
  passwordSecretKeyRef:
    name: passwords
    key: photoprism
  # This field is immutable and defaults to 10
  maxUserConnections: 20
---
apiVersion: v1
kind: Secret
metadata:
  name: passwords
stringData:
  photoprism: photoprismsecret
  root-password: supersecret
```

Logged by MariaDB:
```
2023-04-24 12:23:05 2134 [Warning] Aborted connection 2134 to db: 'unconnected' user: 'unauthenticated' host: '10.244.1.39' (This connection closed normally without authentication)
```

**Environment details**:
- Kubernetes version: v1.26.0
- mariadb-operator version: 0.11.0
- Install method: helm
- Install flavour: default? I haven't customised any helm default values yet.

**Comment by mdbooth** [comment]
The root password is fine. The problem is that during bootstrap we only create `root@localhost`, which is not permitted to connect remotely. This means that the operator cannot connect to the database it has created. I have worked round the problem by setting the following in my `MariaDB` CR:

```yaml
  env:
    - name: MARIADB_ROOT_HOST
      value: '%'
```

Looking at the entrypoint script in the image, this seems to be the only possible way to bootstrap. I looked in the operator, though, and there doesn't seem to be any mention of it. I don't understand how it's possible for mariadb-operator to bootstrap a database without this, because it won't be able to connect. By any chance do the examples need to be updated?

**Comment by mmontes11** [comment]
Thanks for looking into it @mdbooth !

We definetly want to grant `root@%` to the operator so this makes perfect sense. It might be working with some MariDB versions because this is the default value, see:
- https://mariadb.com/kb/en/mariadb-docker-environment-variables/#mariadb_root_host-mysql_root_host

However, as we totally rely on this to be present, it will be worth enforcing this variable and explicitely set the variable here:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/pkg/builder/statefulset_builder.go#L206

> By any chance do the examples need to be updated?

I think it would be worth enforcing this variable internally and not rely on the user for doing this, as I pointed out above.

Would you like to give it a try and contribute to the project? 🙏🏻 🦭

**Comment by mmontes11** [comment]
Hey @mdbooth!

Thanks again for your thorough investigation, pretty useful 🎖️ 

> if you're running the same image as me and not setting this value.

Well, this is nor strictly true, I see you are running `mariadb:10.11.2`,  while I'm using `mariadb:10.7.4` from the examples:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/config/samples/mariadb_v1alpha1_mariadb.yaml#L18

So they might have changed the hacky bash for making the default logic.  

I don't like the fact that the whole operator functionallity relies on a bash hack to be present in certain version of the image, that's not very serious and could potentially lead to errors. It's anyway good to understand what is going on in your case but anyway now I feel even more convinced that we should explicitly add `MARIADB_ROOT_HOST` = `%`:
- https://github.com/mariadb-operator/mariadb-operator/blob/main/pkg/builder/statefulset_builder.go#L206

Thoughts?

**Comment by mmontes11** [comment]
> A few. Counter points first! Firstly given that it's working for you I still think it's likely to be something weird in my environment. I haven't had a chance to debug yet, though, so I have no idea what that might be. Secondly, as you pointed out above the default behaviour which you seem to be getting (but I don't) is documented. Documentation makes an API contract in my view! I'd hope they wouldn't change that lightly.

Alright, in that case we should keep this issue open to keep track of the issue you encounter if related to the operator. Just leave your findings in a comment when you have the chance to debug 👍🏻 

> However, given that we explicitly require the default behaviour I think you're absolutely right and we should make it part of the explicit configuration.

Agreed, I've added the env var in this PR:
- https://github.com/mariadb-operator/mariadb-operator/pull/93

It will be released in the `v0.0.12`, stay tuned!
